File: frontend/src/services/Notification.ts
import api from "../lib/api";
import type { Notification } from "../interfaces/Notification";
import type { User } from "../interfaces/User";

interface RawNotification {
  ID?: number; id?: number;
  title?: string; Title?: string;
  message?: string; Message?: string;
  type?: string; Type?: string;
  user_id?: number; UserID?: number;
  created_at?: string; CreatedAt?: string;
  is_read?: boolean; IsRead?: boolean;
  report_id?: number; ReportID?: number;
  user?: User; User?: User;
}

export async function fetchNotifications(userId: number): Promise<Notification[]> {
  const { data } = await api.get("/notifications", { params: { user_id: userId } });
  const list = (Array.isArray(data) ? data : data?.items || []) as RawNotification[];
  return list.map((n) => ({
    ID: n.ID ?? n.id ?? 0,
    title: n.title ?? n.Title ?? "",
    message: n.message ?? n.Message ?? "",
    type: n.type ?? n.Type ?? "",
    user_id: n.user_id ?? n.UserID ?? 0,
    created_at: (n.created_at ?? n.CreatedAt ?? "")?.toString(),
    is_read: n.is_read ?? n.IsRead ?? false,
    report_id: n.report_id ?? n.ReportID, // ✅ สำคัญ
  })) as Notification[];
}

export async function createNotification(payload: {
  title: string;
  message: string;
  type: string;
  user_id: number;
  report_id?: number;
}): Promise<Notification | null> {
  try {
    const res = await api.post("/notifications", payload);
    return res.data as Notification;
  } catch (err) {
    console.error("❌ createNotification error:", err);
    return null;
  }
}

export async function markNotificationRead(id: number): Promise<void> {
  await api.put(`/notifications/${id}/read`);
}

export async function markAllNotificationsRead(userId: number): Promise<void> {
  await api.put(`/notifications/read-all`, { user_id: userId });
}

export async function deleteNotification(id: number): Promise<void> {
  await api.delete(`/notifications/${id}`);
}

File: backend/controllers/notification.go
// controllers/notification.go
package controllers

import (
	"log"
	"net/http"

	"example.com/sa-gameshop/configs"
	"example.com/sa-gameshop/entity"
	"github.com/gin-gonic/gin"
)

// POST /notifications
func CreateNotification(c *gin.Context) {
        var body entity.Notification
        if err := c.ShouldBindJSON(&body); err != nil {
                c.JSON(http.StatusBadRequest, gin.H{"error": "bad request body"})
                return
        }

	// ตรวจสอบว่ามี user_id นี้จริงไหม
	var user entity.User
	if tx := configs.DB().Where("id = ?", body.UserID).First(&user); tx.RowsAffected == 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "user_id not found"})
		return
	}

       // default type ถ้าไม่ส่งมา
       if body.Type == "" {
               body.Type = "system"
       }
       // เริ่มต้นให้ยังไม่อ่าน
       body.IsRead = false

	if err := configs.DB().Create(&body).Error; err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	log.Printf("📥 CreateNotification: user_id=%d title=%q message=%q type=%q",
		body.UserID, body.Title, body.Message, body.Type)

	c.JSON(http.StatusCreated, body)
}

// GET /notifications?user_id=...
func FindNotifications(c *gin.Context) {
	uid := c.Query("user_id")
	if uid == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "must provide user_id"})
		return
	}

	var rows []entity.Notification
	if err := configs.DB().
		Preload("User").
		Where("user_id = ?", uid).
		Order("created_at DESC").
		Find(&rows).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": err.Error()})
		return
	}

	log.Printf("📤 FindNotifications: user_id=%s count=%d", uid, len(rows))
	c.JSON(http.StatusOK, rows)
}

// GET /notifications/:id
func FindNotificationByID(c *gin.Context) {
	var row entity.Notification
	if tx := configs.DB().Preload("User").First(&row, c.Param("id")); tx.RowsAffected == 0 {
		c.JSON(http.StatusNotFound, gin.H{"error": "notification not found"})
		return
	}
	c.JSON(http.StatusOK, row)
}

// PUT /notifications/:id
func UpdateNotification(c *gin.Context) {
	var payload entity.Notification
	if err := c.ShouldBindJSON(&payload); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	db := configs.DB()
	var row entity.Notification
	if tx := db.First(&row, c.Param("id")); tx.RowsAffected == 0 {
		c.JSON(http.StatusNotFound, gin.H{"error": "notification not found"})
		return
	}

       if err := db.Model(&row).Updates(payload).Error; err != nil {
               c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
               return
       }

       _ = db.First(&row, row.ID)
       c.JSON(http.StatusOK, row)
}

// DELETE /notifications/:id
func DeleteNotificationByID(c *gin.Context) {
	if tx := configs.DB().Exec("DELETE FROM notifications WHERE id = ?", c.Param("id")); tx.RowsAffected == 0 {
		c.JSON(http.StatusNotFound, gin.H{"error": "notification not found"})
		return
	}
	c.JSON(http.StatusOK, gin.H{"message": "deleted successfully"})
}
